File: 2026-01-January-GFG-POTD/January 18 - Next element with greater frequency/Solution.java
class Solution {
    public ArrayList<Integer> nextFreqGreater(int[] arr) {
        int n = arr.length;
        ArrayList<Integer> result = new ArrayList<>();
        
        // Step 1: Frequency map
        HashMap<Integer, Integer> freq = new HashMap<>();
        for (int num : arr) {
            freq.put(num, freq.getOrDefault(num, 0) + 1);
        }
        
        // Step 2: Initialize result with -1
        for (int i = 0; i < n; i++) {
            result.add(-1);
        }
        
        // Step 3: Stack to store indices
        Stack<Integer> stack = new Stack<>();
        
        // Step 4: Traverse array
        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() &&
                   freq.get(arr[i]) > freq.get(arr[stack.peek()])) {
                result.set(stack.pop(), arr[i]);
            }
            stack.push(i);
        }
        
        return result;
    }
}

File: 2026-01-January-GFG-POTD/January 18 - Next element with greater frequency/Solution.cpp
class Solution {
public:
    vector<int> nextFreqGreater(vector<int>& arr) {
        int n = arr.size();
        vector<int> result(n, -1);

        // Step 1: Frequency map
        unordered_map<int, int> freq;
        for (int num : arr) {
            freq[num]++;
        }

        // Step 2: Stack to store indices
        stack<int> st;

        // Step 3: Traverse array
        for (int i = 0; i < n; i++) {
            while (!st.empty() && freq[arr[i]] > freq[arr[st.top()]]) {
                result[st.top()] = arr[i];
                st.pop();
            }
            st.push(i);
        }

        return result;
    }
};
